Added percona mongodb alert rules

yasharne opened the pull request.

paulfantom commented on _data/rules.yml:
```diff
@@ -538,6 +538,50 @@ groups:
+                severity: warning
+              - name: MongoDB too many connections
+                description: Too many connections
+                query: 'mongodb_connections{state="current"} > 500'
```
Why is 500 a condition for an alert? By default mongo allows 64k connections, so alerting on less than 1% of allowed limit is a bit strange.

paulfantom commented on _data/rules.yml:
```diff
@@ -538,6 +538,50 @@ groups:
+                description: MongoDB replication headroom is <= 0
+                query: '(avg(mongodb_mongod_replset_oplog_tail_timestamp - mongodb_mongod_replset_oplog_head_timestamp) - (avg(mongodb_mongod_replset_member_optime_date{state="PRIMARY"}) - avg(mongodb_mongod_replset_member_optime_date{state="SECONDARY"}))) <= 0'
+                severity: critical
+              - name: MongoDB replication Status 3
```
Couldn't this name be more descriptive? "Status 3" doesn't mean much at the first glance.

paulfantom commented on _data/rules.yml:
```diff
@@ -538,6 +538,50 @@ groups:
+                query: '(avg(mongodb_mongod_replset_oplog_tail_timestamp - mongodb_mongod_replset_oplog_head_timestamp) - (avg(mongodb_mongod_replset_member_optime_date{state="PRIMARY"}) - avg(mongodb_mongod_replset_member_optime_date{state="SECONDARY"}))) <= 0'
+                severity: critical
+              - name: MongoDB replication Status 3
+                description: MongoDB Replication set member either perform startup self-checks, or transition from completing a rollback or resync
```
If it is performing automatic action then why immediately wake up human operator (this is what `critical` severity means)?

paulfantom commented on _data/rules.yml:
```diff
@@ -538,6 +538,50 @@ groups:
+                query: "mongodb_mongod_replset_member_state == 3"
+                severity: critical
+              - name: MongoDB replication Status 6
+                description: MongoDB Replication set member as seen from another member of the set, is not yet known
```
If it is down, does it mean cluster is not operational? If cluster is operational, then `critical` severity is a bit too much.

paulfantom commented on _data/rules.yml:
```diff
@@ -538,6 +538,50 @@ groups:
+              - name: MongoDB replication Status 8
+                description: MongoDB Replication set member as seen from another member of the set, is unreachable
+                query: "mongodb_mongod_replset_member_state == 8"
+                severity: critical
```
If it is unreachable, does it mean cluster is not operational? If cluster is operational, then critical severity is a bit too much.

paulfantom commented on _data/rules.yml:
```diff
@@ -538,6 +538,50 @@ groups:
+                query: "mongodb_mongod_replset_member_state == 8"
+                severity: critical
+              - name: MongoDB replication Status 9
+                description: MongoDB Replication set member is actively performing a rollback. Data is not available for reads
```
Is data not available from whole cluster or only from one node? Does it effectively mean "cluster is down" when looking from client perspective? If not, then `critical` severity is too much.

paulfantom commented on _data/rules.yml:
```diff
@@ -538,6 +538,50 @@ groups:
+                query: "mongodb_mongod_replset_member_state == 9"
+                severity: critical
+              - name: MongoDB replication Status 10
+                description: MongoDB Replication set member was once in a replica set but was subsequently removed
```
Why is this `critical`?

yasharne commented:
I have added `mondodb instance down` alert and changed the `too many
connections` alert to fire when the connections are more than 80% of the
available connections and removed `mongodb_replset_member_state` based alerts as I don't have
enough information on them

samber commented on _data/rules.yml:
```diff
@@ -538,6 +538,34 @@ groups:
+                severity: warning
+              - name: MongoDB too many connections
+                description: Too many connections
+                query: 'avg by(instance) (max_over_time(mongodb_connections{state="current"}[5m])) / avg by(instance) (sum (mongodb_connections) by (instance)) * 100 > 80'
```
I don't have a running mongodb exporter in my infrastructure. Could please explain the `mongodb_connections` value and labels?

Are you sure this metric is part of Percona exporter?
It looks very similar to dcu's.

yasharne replied:
Yes, I'm using this percona exporter for our mongodb servers, and the exported labels are `instance`, `job` and `state` (which is either `current` or `available`)

samber commented on _data/rules.yml:
```diff
@@ -538,6 +538,34 @@ groups:
+                severity: warning
+              - name: MongoDB virtual memory usage
+                description: High memory usage
+                query: '(sum(mongodb_memory{type="virtual"}) BY (instance) / sum(mongodb_memory{type="mapped"}) BY (instance)) > 3'
```
Are you sure this metric is part of Percona exporter?
It looks very similar to dcu's.

yasharne replied:
Yes, these metrics are also exported by Percona exporter

samber reviewed (approved):
Thanks @yasharne for contributing.

I just added a few comments.

samber reviewed (changes requested):
oops

yasharne commented on _data/rules.yml:
```diff
@@ -538,6 +538,34 @@ groups:
+                severity: critical
+              - name: MongoDB replication headroom
+                description: MongoDB replication headroom is <= 0
+                query: '(avg(mongodb_mongod_replset_oplog_tail_timestamp - mongodb_mongod_replset_oplog_head_timestamp) - (avg(mongodb_mongod_replset_member_optime_date{state="PRIMARY"}) - avg(mongodb_mongod_replset_member_optime_date{state="SECONDARY"}))) <= 0'
```
`mongodb_mongod_replset_oplog_tail_timestamp` and `mongodb_mongod_replset_oplog_head_timestamp` doesn't have any label about their replica set, so I can only aggregate them by their instance

vfiset replied:
Hi @samber @yasharne I was about to open a PR but find out this one which is more or less related. 

My PR was about removing the alerts based on mongodb_mongod_replset_oplog* since they have been [deprecated](https://github.com/percona/mongodb_exporter#important-note).

Will still proceed with my PR but it think it explains the lack of replica set label

yasharne commented:
Thanks @samber for your feedback, I made few changes but I have commented on one of your comments

samber commented:
Ok for me, let's merge ;)
